test/gtest/uct/test_peer_failure.cc

```diff
@@ -538,3 +540,59 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure_multiple, test,
 }
 
 UCT_INSTANTIATE_TEST_CASE(test_uct_peer_failure_multiple)
+
+class test_uct_peer_failure_keepalive : public test_uct_peer_failure
+{
+public:
+
+protected:
+    void is_skipped()
+    {
+        if (!has_transport("rc_mlx5")) {
+            UCS_TEST_SKIP_R("Unsupported");
+        }
+    }
+};
+
+UCS_TEST_P(test_uct_peer_failure_keepalive, killed)
+{
+    ucs_status_t status;
+
+    is_skipped();
+
+    scoped_log_handler slh(wrap_errors_logger);
+    flush();
+    kill_receiver();
+
+    status = uct_ep_flush(ep0(), UCT_FLUSH_FLAG_KEEPALIVE, NULL);
+    ASSERT_UCS_OK(status);
+    flush();
+
+    EXPECT_EQ(m_err_count, 1);
+}
+
+UCS_TEST_P(test_uct_peer_failure_keepalive, resources)
+{
+    ucs_status_t status;
+
+    is_skipped();
+
+    scoped_log_handler slh(wrap_errors_logger);
+
+    /* send dummy data, but don't progress iface - it will leave
+     * resource counters untouched and flush(KEEPALIVE) should
+     * not make anything */
+    send_am(0);
+    kill_receiver();
+    /* after kill_receiver all send/recv operations will fail. flush(KEEPALIVE)
+     * should not make anything and no errors should be registered */
+
+    status = uct_ep_flush(ep0(), UCT_FLUSH_FLAG_KEEPALIVE, NULL);
+    ASSERT_UCS_OK(status);
+    flush();
+
+    EXPECT_EQ(m_err_count, 0);
```

Comment:
1. reverse arg order
2. why we expect no errors? receiver was killed

Reply:
doesn't this mean a problem with flush (CHECK)? i'd expect it to indicate when error hapens

Reply:
no. here used some kind of controlled race condition: we are emulating state when peer was killed just after HW received data & ack was sent. In real life peer failure will be detected on next keepalive iteration